src/processor.jl

```diff
@@ -271,3 +273,38 @@ Write a log event
 function write_event(ctx::Context, event::Event)
     write_event(ctx.log_sink, event)
 end
+
+"""
+    lock(f, ctx::Context)
+
+Acquire `ctx.proc_lock`, execute `f` with the lock held, and release the lock when `f` returns.
+"""
+Base.lock(f, ctx::Context) = lock(f, ctx.proc_lock)
```

Comment:
This also needs to be applied to `procs(ctx)` for correctness:

```julia
procs(ctx) = lock(ctx) do
    copy(ctx.procs)
end
```

Reply:
Indeed it does. Requires a bit of a rewrite of `compute_dag` but nothing too serious.